src/workerd/io/trace.h

```diff
@@ -109,13 +109,114 @@ class TraceId final {
     return high;
   }
 
+  static TraceId fromCapnp(rpc::InvocationSpanContext::TraceId::Reader reader);
+  void toCapnp(rpc::InvocationSpanContext::TraceId::Builder writer) const;
+
 private:
   uint64_t low = 0;
   uint64_t high = 0;
 };
 constexpr TraceId TraceId::nullId = nullptr;
 
+// The InvocationSpanContext is a tuple of a trace id, invocation id, and span id.
+// The trace id represents a top-level request and should be shared across all
+// invocation spans and events within those spans. The invocation id identifies
+// a specific worker invocation. The span id identifies a specific span within an
+// invocation. Every invocation of every worker should have an InvocationSpanContext.
+// That may or may not have a trigger InvocationSpanContext.
+class InvocationSpanContext final: public kj::Refcounted,
+                                   public kj::EnableAddRefToThis<InvocationSpanContext> {
+public:
+  // Spans within a InvocationSpanContext are identified by a span id that is a
+  // monotically increasing number. Every InvocationSpanContext has a root span
+  // whose ID is zero. Every child span context created within that context will
+  // have a span id that is one greater than the previously created one.
+  class SpanIdCounter final: public kj::Refcounted {
+  public:
+    SpanIdCounter() = default;
+    KJ_DISALLOW_COPY_AND_MOVE(SpanIdCounter);
+
+    inline kj::uint next() {
+      static constexpr kj::uint kMax = kj::maxValue;
+      KJ_ASSERT(id < kMax, "max number of spans exceeded");
```

Comment:
Not objecting to this assert, but if 2^32 spans are reached we have bigger problems 😄

Reply:
Oh, I meant for this to be a `KJ_DASSERT` but this is just me being way over paranoid lol.